packages/@expo/cli/src/export/exportAssets.ts

```diff
@@ -12,26 +12,58 @@ import { uniqBy } from '../utils/array';
 
 const debug = require('debug')('expo:export:exportAssets') as typeof console.log;
 
+function mapAssetHashToAssetString(asset: Asset, hash: string) {
+  return 'asset_' + hash + ('type' in asset && asset.type ? '.' + asset.type : '');
+}
+
+export function assetPatternsToBeBundled(
+  exp: ExpoConfig & { extra?: { assetPatternsToBeBundled?: string[] } }
+): string[] | undefined {
+  return exp?.extra?.assetPatternsToBeBundled?.length
+    ? exp?.extra?.assetPatternsToBeBundled
+    : undefined;
+}
+
 /**
- * Resolves the assetBundlePatterns from the manifest and returns a list of assets to bundle.
- *
- * @modifies {exp}
+ * Given an asset and a set of strings representing the assets to be bundled, returns true if
+ * the asset is part of the set to be bundled.
+ * @param asset Asset object
+ * @param bundledAssetsSet Set of strings
+ * @returns true if the asset should be bundled
  */
-export async function resolveAssetBundlePatternsAsync<T extends ExpoConfig>(
-  projectRoot: string,
-  exp: T,
-  assets: Asset[]
-): Promise<Omit<T, 'assetBundlePatterns'> & { bundledAssets?: string[] }> {
-  if (!exp.assetBundlePatterns?.length || !assets.length) {
-    delete exp.assetBundlePatterns;
-    return exp;
+function assetShouldBeIncludedInExport(asset: Asset, bundledAssetsSet: Set<string> | undefined) {
+  if (!bundledAssetsSet) {
+    return true;
   }
+  return (
+    asset.fileHashes.filter((hash) => bundledAssetsSet?.has(mapAssetHashToAssetString(asset, hash)))
+      .length > 0 ?? false
+  );
+}
+
+/**
+ * Computes a set of strings representing the assets to be bundled with an export, given an array of assets,
+ * and a set of patterns to match
+ * @param assets The asset array
+ * @param assetPatternsToBeBundled An array of strings with glob patterns to match
+ * @param projectRoot The project root
+ * @returns A set of asset strings
+ */
+function setOfAssetsToBeBundled(
+  assets: Asset[],
+  assetPatternsToBeBundled: string[] | undefined,
```

Comment:
this is always called with this arg defined

```suggestion
  assetPatternsToBeBundled: string[],
```